crates/prune/src/segments/history.rs

```diff
@@ -34,79 +33,180 @@ where
+            cursor.seek_exact(last_key(&key))?;
+        }
+    } else {
+        trace!(target: "pruner::history",
+            to_block,
+            highest_block_shard=key.as_ref().highest_block_number,
+            "cannot prune complete shard"
+        );
+
         // Shard contains block numbers that are higher than the target one, so we need to
         // filter it. It is guaranteed that further shards for this sharded key will not
         // contain the target block number, as it's in this shard.
-        else {
-            let new_blocks =
-                blocks.iter().skip_while(|block| *block <= to_block).collect::<Vec<_>>();
-
-            // If there were blocks less than or equal to the target one
-            // (so the shard has changed), update the shard.
-            if blocks.len() as usize != new_blocks.len() {
-                // If there are no more blocks in this shard, we need to remove it, as empty
-                // shards are not allowed.
-                if new_blocks.is_empty() {
-                    if key.as_ref().highest_block_number == u64::MAX {
-                        let prev_row = cursor.prev()?;
-                        match prev_row {
-                            // If current shard is the last shard for the sharded key that
-                            // has previous shards, replace it with the previous shard.
-                            Some((prev_key, prev_value)) if key_matches(&prev_key, &key) => {
-                                cursor.delete_current()?;
-                                deleted += 1;
-                                // Upsert will replace the last shard for this sharded key with
-                                // the previous value.
-                                cursor.upsert(key.clone(), prev_value)?;
-                            }
-                            // If there's no previous shard for this sharded key,
-                            // just delete last shard completely.
-                            _ => {
-                                // If we successfully moved the cursor to a previous row,
-                                // jump to the original last shard.
-                                if prev_row.is_some() {
-                                    cursor.next()?;
-                                }
-                                // Delete shard.
-                                cursor.delete_current()?;
-                                deleted += 1;
+        let old_blocks = blocks.iter().skip_while(|block| *block <= to_block).collect::<Vec<_>>();
+
+        // If there were blocks less than or equal to the target one
+        // (so the shard has changed), update the shard.
+        if blocks.len() as usize != old_blocks.len() {
+            // If there are no more blocks in this shard, we need to remove it, as empty
+            // shards are not allowed.
+            if old_blocks.is_empty() {
```

Comment:
this seems like a bug, there is still blocks in the shard, since otherwise the check `if blocks.len() as usize != old_blocks.len()` wouldn't pass @shekhirin will open an issue

Reply:
why? if `blocks = [1, 2, 3]`, `to_block = 3`, then `old_blocks = []` and the check will pass `blocks.len() != old_blocks.len()`. Then, we'll go into this clause `old_blocks.is_empty()` and remove the shard completely.

Reply:
https://github.com/paradigmxyz/reth/issues/7104

Reply:
yeah exactly, so there are still blocks `[1, 2, 3]` in the shard. old blocks helped me think about it: lower block number = older block

Reply:
shard should only be completely removed if there is guarantee all blocks are lower than target block. which is already done here

https://github.com/paradigmxyz/reth/blob/e4f34e79ab144734d38661b66e3ef3e32b9a154d/crates/prune/src/segments/history.rs#L40-L42

Reply:
renamed `old_blocks` to `lower_blocks`, to stick to standard concept of 'block height'

Reply:
fudge, `higher_blocks` it should be

Reply:
sorry, I needed to clarify my example. If it's **the last shard** for this sharded key, then the `if key.as_ref().highest_block_number <= to_block {` check will not pass, because the `highest_block_number == u64::MAX`. So we go into this clause, filter the blocks and realize the shard is empty now, and delete it.

But I think this clause https://github.com/paradigmxyz/reth/blob/d6aea5f99da9ff85aaaddc8a426313123929bd7c/crates/prune/src/segments/history.rs#L167-L176 is actually redundant, it shouldn't be possible to get into it. I think I put it here jus to be sure all if-else branches are considered.